### PR TITLE
[Terra-Drop-Down-Button] add navigation between last and first items

### PR DESCRIPTION
**Pull request by AH106586Harika**
### Summary
[UXPLATFORM-8764]: (https://jira2.cerner.com/browse/UXPLATFORM-8764)Terra-Dropdown- Enable selection from first Item to last Item using keyboard navigation

### Solution
Added logic to change the selection when we are on last option to first option and vice-versa

### CODE REVIEW

**Comment by sugan2416** [comment]
@AH106586Harika  Please update the changelog for `terra-dropdown-button` (packages/terra-dropdown-button/CHANGELOG.md) and mark pr for review

**Comment by sugan2416** [comment]
can we update the tests for first and last item focus?

File: packages/terra-dropdown-button/CHANGELOG.md [hunk 1]
```diff
@@ -1,6 +1,8 @@
 # Changelog
 
 ## Unreleased
+* Fixed
+   * Added logic to change the selection when we are on last option to first option and vice-versa
```
Comment (sdadn) [inline packages/terra-dropdown-button/CHANGELOG.md:5]:
```suggestion

* Changed
   * Updated selection logic to cycle from last option to first option and vice-versa
```

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 2]
```diff
@@ -268,3 +268,23 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
     Terra.validates.element('variants');
   });
 });
+
+describe('Navigate from first item to last item', () => {
```
Comment (sdadn) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:272]:
```suggestion
describe('when navigating between first and last items', () => {
```

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 3]
```diff
@@ -268,3 +268,23 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
...
+  it('Navigate from first Item to last Item', () => {
+    browser.url('/raw/tests/cerner-terra-core-docs/dropdown-button/variant-dropdown-buttons');
+    $('[class*=dropdown-button]').click();
+    $('[class*=dropdown-list]').waitForDisplayed();
+    $('li:first-child[role="menuitem"]').isFocused();
+    browser.keys(['ArrowUp']);
+    Terra.validates.element('last item focused', { selector: '#root' });
+  });
+
+  it('Navigate from last item to first item', () => {
```
Comment (sdadn) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:282]:
```suggestion
  it('Navigates from last item to first item', () => {
```

File: packages/terra-dropdown-button/src/_DropdownList.jsx [hunk 4]
```diff
@@ -63,12 +64,20 @@ class DropdownList extends React.Component {
...
       }
       event.preventDefault();
     } else if (keyCode === KeyCode.KEY_UP) {
       if (!this.pressed) {
-        this.changeFocusState(Util.findPrevious(this, this.state.focused));
+        if (index === 0) {
+          this.changeFocusState(Util.getChildArray(this).length - 1);
+        } else {
+          this.changeFocusState(Util.findPrevious(this, this.state.focused));
+        }
```
Comment (sdadn) [inline packages/terra-dropdown-button/src/_DropdownList.jsx:80]: Can this also be tested through Jest unit tests?

File: packages/terra-dropdown-button/CHANGELOG.md (hunk 1 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/CHANGELOG.md]: Done.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 2 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js]: Done.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 5]
```diff
@@ -268,3 +268,23 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
     Terra.validates.element('variants');
   });
 });
+
+describe('Navigate from first item to last item', () => {
+  it('Navigate from first Item to last Item', () => {
```
[the remark this answers is not in the document]
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js]: Done.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 3 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js]: Done.

File: packages/terra-dropdown-button/src/_DropdownList.jsx (hunk 4 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:80]: This scenario has already been tested with WDIO since it is more related to the function like how the focus is set on the keydown. WDIO was more suitable for this as it provides snapshot evidence of the item being focused rather than validating on the Jest.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 6]
```diff
@@ -109,20 +109,20 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
...
       });
 
-      it('keyboard navigates down past the last option', () => {
-        browser.keys(['ArrowDown', 'ArrowDown', 'ArrowDown', 'ArrowDown', 'Space']);
-        Terra.validates.element('no down out of bounds');
-      });
+      // it('keyboard navigates down past the last option', () => {
+      //   browser.keys(['ArrowDown', 'ArrowDown', 'ArrowDown', 'ArrowDown', 'Space']);
+      //   Terra.validates.element('no down out of bounds');
+      // });
```
Comment (sycombs) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:115]: Are these tests no longer needed, and if so can they be deleted?

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 7]
```diff
@@ -267,4 +267,24 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
 
     Terra.validates.element('variants');
   });
+
+  describe('When navigating between first and last items', () => {
+    it('Navigates from first item to last item', () => {
+      browser.url('/raw/tests/cerner-terra-core-docs/dropdown-button/variant-dropdown-buttons');
+      $('[class*=dropdown-button]').click();
+      $('[class*=dropdown-list]').waitForDisplayed();
+      $('li:first-child[role="menuitem"]').isFocused();
```
Comment (sycombs) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:276]: This line is not doing anything, it is simply returning a boolean without any operation (see [the doc](https://webdriver.io/docs/api/element/isFocused/) for isFocused); can we update this to use an `expect` similar to [this](https://github.com/cerner/terra-core/blob/35cbecb6941c298631f9eabd0dd3fe1795622a55/packages/terra-show-hide/tests/wdio/show-hide-spec.js#L25)?

File: packages/terra-dropdown-button/src/_DropdownList.jsx (hunk 4 above)
Comment (sycombs) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:80]: Can we add some tests for other/edge cases (can be Jest or WDIO) such as:

- index != 0
- There are only 1 or 2 items in the dropdown
- The list is empty

**Review by sycombs — CHANGES_REQUESTED** [review]

File: packages/terra-dropdown-button/src/_DropdownList.jsx (hunk 4 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:80]: https://user-images.githubusercontent.com/126655256/233158147-273b4949-d654-41aa-953c-ae65a9eea496.mp4

Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:80]: https://user-images.githubusercontent.com/126655256/233160783-4f6c530f-f490-402c-b231-b8e377f9db87.mp4

Comment (sycombs) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:80]: Thanks for adding the screen recording! Can we also add a WDIO test for the 1 dropdown item case?

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 6 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js]: Deleted.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 7 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js]: Reverted these changes.

File: packages/terra-dropdown-button/src/_DropdownList.jsx (hunk 4 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/src/_DropdownList.jsx:80]: Added testcase for 1 dropdown item

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 8]
```diff
@@ -109,19 +109,19 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
         Terra.validates.element('click');
       });
 
-      it('keyboard navigates down past the last option', () => {
-        browser.keys(['ArrowDown', 'ArrowDown', 'ArrowDown', 'ArrowDown', 'Space']);
-        Terra.validates.element('no down out of bounds');
+      it('Navigates from last item to first item', () => {
+        browser.keys(['ArrowDown', 'ArrowDown', 'ArrowDown', 'ArrowDown', 'Enter']);
+        Terra.validates.element('first item focused');
```
Comment (supreethmr) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:114]: I still see old screenshots `no down out of bounds` for both dropdown-button on master repo which are supposed to be removed with this update

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js [hunk 9]
```diff
@@ -259,6 +259,13 @@ Terra.describeViewports('Dropdown Button', ['medium'], () => {
       $('#first').click();
       Terra.validates.element('first button clicked');
     });
+
+    it('Focus to single item on arrow up and arrow down', () => {
+      $('[class*=dropdown-button]').click();
+      $('[class*=dropdown-list]').waitForDisplayed();
+      browser.keys(['ArrowDown', 'ArrowDown', 'Enter']);
```
Comment (supreethmr) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:266]:
```suggestion
      browser.keys(['ArrowDown', 'ArrowDown',]);
```
I think the test was added to ensure multiple ArrowDown was not thrown any error when list has just one item !! I

File: packages/terra-dropdown-button/tests/wdio/split-button-spec.js [hunk 10]
```diff
@@ -159,19 +159,19 @@ Terra.describeViewports('Split Button', ['medium'], () => {
...
         browser.keys(['ArrowDown', 'ArrowDown', 'ArrowUp', 'Enter']);
         Terra.validates.element('up arrow');
       });
 
-      it('keyboard navigates up past the first option', () => {
-        browser.keys(['ArrowDown', 'ArrowDown', 'ArrowUp', 'ArrowUp', 'Enter']);
-        Terra.validates.element('no up out of bounds');
+      it('Navigates from first item to last item', () => {
+        browser.keys(['ArrowUp', 'Enter']);
+        Terra.validates.element('last item focused');
```
Comment (supreethmr) [inline packages/terra-dropdown-button/tests/wdio/split-button-spec.js:174]: Same here  you need to remove old snapshots

File: packages/terra-dropdown-button/tests/wdio/split-button-spec.js [hunk 11]
```diff
@@ -319,6 +319,13 @@ Terra.describeViewports('Split Button', ['medium'], () => {
       $('#second').click();
       Terra.validates.element('second button clicked');
     });
+
+    it('Focus to single item on arrow up and arrow down', () => {
+      $('[class*=split-button-caret]').click();
+      $('[class*=dropdown-list]').waitForDisplayed();
+      browser.keys(['ArrowDown', 'ArrowDown', 'Enter']);
```
Comment (supreethmr) [inline packages/terra-dropdown-button/tests/wdio/split-button-spec.js:326]:
```suggestion
      browser.keys(['ArrowDown', 'ArrowDown']);
```

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 8 above)
Comment (sdadn) [inline packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:114]: Can we also test for focus using `expect`? It will be something like this (replace " first list item" with the appropriate selector):
```suggestion
        expect($(['first list item']).isFocused()).toBeTruthy();
        Terra.validates.element('first item focused');
```

**Review by sdadn — CHANGES_REQUESTED** [review]

File: packages/terra-dropdown-button/tests/wdio/split-button-spec.js [hunk 12]
```diff
@@ -159,19 +159,19 @@ Terra.describeViewports('Split Button', ['medium'], () => {
         Terra.validates.element('click');
       });
 
-      it('keyboard navigates down past the last option', () => {
-        browser.keys(['ArrowDown', 'ArrowDown', 'ArrowDown', 'ArrowDown', 'Space']);
-        Terra.validates.element('no down out of bounds');
+      it('Navigates from last item to first item', () => {
+        browser.keys(['ArrowDown', 'ArrowDown', 'ArrowDown', 'ArrowDown', 'Enter']);
+        Terra.validates.element('first item focused');
```
Comment (sdadn) [inline packages/terra-dropdown-button/tests/wdio/split-button-spec.js:164]: Same here about the focus test:

**Review by sdadn — CHANGES_REQUESTED** [review]

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 8 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:114]: Removed old screenshots.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 9 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js]: Fixed.

File: packages/terra-dropdown-button/tests/wdio/split-button-spec.js (hunk 11 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/split-button-spec.js]: Done.

File: packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js (hunk 8 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/dropdown-button-spec.js:114]: Added test case to focus first list item.

File: packages/terra-dropdown-button/tests/wdio/split-button-spec.js (hunk 12 above)
Comment (AH106586Harika) [reply packages/terra-dropdown-button/tests/wdio/split-button-spec.js:164]: Added test case to focus first list item.

**Review by sycombs — APPROVED** [review]

**Review by supreethmr — APPROVED** [review]

File: packages/terra-dropdown-button/CHANGELOG.md [hunk 13]
```diff
@@ -1,6 +1,8 @@
 # Changelog
 
 ## Unreleased
+* Fixed
```
Comment (supreethmr) [inline packages/terra-dropdown-button/CHANGELOG.md:4]: nit : 
```suggestion

* Fixed
```

**Review by saket2403 — APPROVED** [review]
